### PR TITLE
Support generation of shared libraries in Windows

### PR DESCRIPTION
**Pull request by traversaro**
Additional related modifications:
  * Cleanup of visibility macros (we just use default visibility on *nix and `CMAKE_EXPORT_ALL_SYMBOLS` in Windows)
  * Add possibility of selecting if the library is build as shared or static using `BUILD_SHARED_LIBS`

### CODE REVIEW

**Comment by traversaro** [comment]
AppVeyor builds are failing for the following reason: 
* `[[noreturn]]` does not work on VS2013 
* Tests are not working on all other builds

**Comment by traversaro** [comment]
The two problems were the usual problem of dll not found by test in Windows (see https://github.com/robotology/how-to-export-cpp-library/issues/4 for a more in-depth discussion) and the fact that I removed the `__declspec(dllexport)` even from the `MODULE` that is meant to be loaded at runtime. 
`CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` remove the need of exporting
symbols in library linked at link time, but dllexport
is still need for function loaded at runtime .

File: CMakeLists.txt [hunk 1]
```diff
@@ -5,7 +5,7 @@
 #  Copy Policy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
 #
 
-cmake_minimum_required(VERSION 3.1)
+cmake_minimum_required(VERSION 3.5)
```
Comment (Nicogene) [inline CMakeLists.txt]: Isn't too much 3.5 as cmake minimum required?

File: doc/release/v1_4_0.md [hunk 2]
```diff
@@ -13,6 +13,11 @@ Important Changes
   compatibility.
 * Added `target_include_directories` in all RTF
   libraries.
+* The RTF libraries library can now be compiled either as shared or
```
Comment (Nicogene) [inline doc/release/v1_4_0.md]: If we decide to set cmake_minimum_required to 3.5 we have to put this information in Important changes in the changelog

Comment (traversaro) [reply doc/release/v1_4_0.md]: Indeed, adding now.

File: CMakeLists.txt (hunk 1 above)
Comment (traversaro) [reply CMakeLists.txt]: `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` requires CMake 3.4 . We can differentiate between different operating systems, but given that most of the time we are using Ubuntu Xenial with CMake 3.5 and CMake 3.5 or greater is supported (directly or using backports) in all system that we support (see https://github.com/robotology/yarp/pull/1350#issuecomment-331796555 ) I think it is not realistic to claim support for CMake < 3.5 if we are not going to test it anywhere.

File: doc/release/v1_4_0.md (hunk 2 above)
Comment (traversaro) [reply doc/release/v1_4_0.md]: Fixed in https://github.com/robotology/robot-testing/pull/85/commits/25fbf265ec27504c89d172ddadb7407e6393ce1f .

**Comment by aerydna** [comment]
@traversaro if you resolve the conflicts we can merge it.. 
btw we have to think a method to add feature notes to the changelog file without everyone operating directly on it. something like this

https://github.com/masukomi/changelog_manager

**Comment by aerydna** [comment]
merged, thanks!

**Pull request merged**